crates/p2p/src/swarm/mod.rs

```diff
@@ -536,58 +548,249 @@ where
         Ok(())
     }
 
-    #[instrument(skip(self, _event))]
-    async fn handle_response_request_event(
+    #[instrument(skip(self, event))]
+    async fn handle_request_response_event(
+        &mut self,
+        event: RequestResponseEvent<GetMessageRequest, GetMessageResponse>,
+    ) -> P2PResult<()> {
+        let RequestResponseEvent::Message { peer, message } = event else {
+            return Ok(());
+        };
+
+        match message {
+            request_response::Message::Request {
+                request_id,
+                request,
+                channel,
+            } => {
+                let Some(req) = v1::GetMessageRequest::from_msg(request) else {
+                    debug!(%peer, "Peer sent invalid get message request, disconnecting it");
+                    let _ = self.swarm.disconnect_peer_id(peer);
+                    return Ok(());
+                };
+
+                let msg = self.handle_get_message_request(req).await?;
+
+                if msg.is_none() {
+                    debug!(%request_id, "Have no needed data, requesting from neighbours");
+                    return Ok(()); // TODO(NikitaMasych): launch recursive request.
+                }
+
+                let response = GetMessageResponse {
+                    msg: vec![msg.unwrap()],
+                };
+
+                let _ = self
+                    .swarm
+                    .behaviour_mut()
+                    .request_response
+                    .send_response(channel, response)
+                    .inspect_err(|_| debug!("Failed to send response"));
+            }
+
+            request_response::Message::Response {
+                request_id,
+                response,
+            } => {
+                if response.msg.is_empty() {
+                    debug!(%request_id, "Have no needed data, requesting from neighbours");
+                    return Ok(()); // TODO(NikitaMasych): launch recursive request.
```

Comment:
I think the node should request this data again after some time instead, so we may have `TimeoutManager` to be an event-loop, reactor kind of thing instead of a place for timeouts only. But let's consider this in next PR